src/NodaTime/Globalization/NodaFormatInfo.cs

```diff
@@ -388,22 +417,34 @@ internal static NodaFormatInfo GetFormatInfo([NotNull] CultureInfo cultureInfo)
 
         /// <summary>
         /// Gets the <see cref="NodaFormatInfo" /> for the given <see cref="IFormatProvider" />. If the
-        /// format provider is null or if it does not provide a <see cref="NodaFormatInfo" />
-        /// object then the format object for the current thread is returned.
+        /// format provider is null then the format object for the current thread is returned. If it's
+        /// a CultureInfo, that's used for everything. If it's a DateTimeFormatInfo, that's used for
+        /// format strings, day names etc but the invariant culture is used for text comparisons and
+        /// resource lookups. Otherwise, <see cref="ArgumentException"/> is thrown.
         /// </summary>
         /// <param name="provider">The <see cref="IFormatProvider" />.</param>
+        /// <exception cref="ArgumentException">The format provider cannot be used for Noda Time.</exception>
         /// <returns>The <see cref="NodaFormatInfo" />. Will never be null.</returns>
         public static NodaFormatInfo GetInstance(IFormatProvider provider)
         {
-            if (provider != null)
+            // TODO: Use C# 7 pattern matching for this...
+            if (provider == null)
             {
-                var cultureInfo = provider as CultureInfo;
-                if (cultureInfo != null)
-                {
-                    return GetFormatInfo(cultureInfo);
-                }
+                return GetInstance(CultureInfo.CurrentCulture);
```

Comment:
Not simply `provider = CultureInfo.CurrentCulture;` and allow the function to continue?

(On the other hand, I do quite like the delegation, so you could keep it too.)